api/src/runtime.rs

```diff
@@ -28,90 +40,178 @@ pub fn bootstrap(
+                tls_cert_path
+            ))?;
+            let key = std::fs::read_to_string(tls_key_path).context(format!(
+                "Failed to read TLS key from path: {}",
+                tls_key_path
+            ))?;
+            let rustls_certificate = RustlsCertificate::new().cert(cert).key(key);
+            let rustls_config = RustlsConfig::new().fallback(rustls_certificate);
+            TcpListener::bind(address).rustls(rustls_config).boxed()
         }
-    }
+        _ => {
+            info!("Not using TLS for API");
+            TcpListener::bind(address).boxed()
+        }
+    };
+
+    let acceptor = tokio::task::block_in_place(move || {
+        runtime_handle
+            .block_on(async move { listener.into_acceptor().await })
+            .with_context(|| format!("Failed to bind Poem to address: {}", address))
+    })?;
+
+    let actual_address = &acceptor.local_addr()[0];
+    let actual_address = *actual_address
+        .as_socket_addr()
+        .context("Failed to get socket addr from local addr for Poem webserver")?;
+    runtime_handle.spawn(async move {
+        let cors = Cors::new()
+            .allow_methods(vec![Method::GET, Method::POST])
+            .allow_headers(vec![header::CONTENT_TYPE, header::ACCEPT]);
+        let route = Route::new()
+            .nest(
+                "/v1",
+                Route::new()
+                    .nest("/", api_service)
+                    .at("/spec.json", spec_json)
+                    .at("/spec.yaml", spec_yaml)
+                    // TODO: We add this manually outside of the OpenAPI spec for now.
+                    // https://github.com/poem-web/poem/issues/364
+                    .at(
+                        "/set_failpoint",
+                        poem::get(set_failpoints::set_failpoint_poem).data(context.clone()),
+                    ),
+            )
+            .with(cors)
+            .with(PostSizeLimit::new(size_limit))
+            // NOTE: Make sure to keep this after all the `with` middleware.
+            .catch_all_error(convert_error)
+            .around(middleware_log);
+        Server::new_with_acceptor(acceptor)
+            .run(route)
+            .await
+            .map_err(anyhow::Error::msg)
+    });
+
+    info!(
+        "Poem is running at {}, behind the reverse proxy at the API port",
+        actual_address
+    );
+
+    Ok(actual_address)
 }
 
 #[cfg(test)]
 mod tests {
     use std::time::Duration;
 
+    use aptos_api_test_context::{new_test_context, TestContext};
     use aptos_config::config::NodeConfig;
     use aptos_types::chain_id::ChainId;
 
-    use crate::{
-        runtime::bootstrap,
-        tests::{new_test_context, TestContext},
-    };
+    use super::bootstrap;
 
+    // TODO: Unignore this when I figure out why this only works when being
+    // run alone (it fails when run with other tests).
+    // https://github.com/aptos-labs/aptos-core/issues/2977
+    #[ignore]
```

Comment:
This likely has some shared state that's an issue

Reply:
yeah, not worth fixing right now